### PR TITLE
BUG: GeoDataFrame constructor set `copy=true` if copy is not set and `data` is a `DataFrame`

### PR DESCRIPTION
**Pull request by spolloni**
I accidentally deleted my geopandas fork so this is just https://github.com/geopandas/geopandas/pull/2306 born again.

**_Original PR description:_** 

This PR is very similar to https://github.com/geopandas/geopandas/pull/1324 and fixes https://github.com/geopandas/geopandas/issues/2301 (as well as duplicates https://github.com/geopandas/geopandas/issues/1804, https://github.com/geopandas/geopandas/issues/2017, https://github.com/geopandas/geopandas/issues/2186, and https://github.com/geopandas/geopandas/issues/2301).

In the spirit of [this logic](https://github.com/pandas-dev/pandas/blob/ab42f85f192ab054a18f94825ced1bb4c1ab7d3f/pandas/core/frame.py#L616-L629), if a DataFrame is passed to the GeoDataFrame constructor, and the copy kwarg is not explicitly set by the user, then copy=True is set so that the input is not modified.

### CODE REVIEW

**Comment by martinfleis** [comment]
This looks okay to me. Can you also add a changelog note?

File: geopandas/tests/test_geodataframe.py [hunk 1]
```diff
@@ -772,6 +772,27 @@ def test_dataframe_to_geodataframe(self):
...
+            df,
+            geometry=points_from_xy(df["longitude"], df["latitude"]),
+            crs=self.df.crs,
+        )
+        assert isinstance(df, pd.DataFrame)
+        assert "geometry" not in df
+        pd.testing.assert_frame_equal(df, df_copy)
+        assert isinstance(gf, GeoDataFrame)
+        assert hasattr(gf, "geometry")
+
```
Comment (m-richards) [inline geopandas/tests/test_geodataframe.py:795]: Perhaps overkill, but I'm wondering if it also makes sense to explicitly check that  mutating the columns in gf doesn't update df:
```suggestion
gf.loc[0, "A" ] =7
pd.testing.assert_frame_equal(df, df_copy)
gf["A"] =3
pd.testing.assert_frame_equal(df, df_copy)
```

File: geopandas/tests/test_geodataframe.py [hunk 2]
```diff
@@ -772,6 +772,27 @@ def test_dataframe_to_geodataframe(self):
...
+            },
+            index=self.df.index,
+        )
+        df_copy = df.copy()
+        gf = GeoDataFrame(
+            df,
+            geometry=points_from_xy(df["longitude"], df["latitude"]),
+            crs=self.df.crs,
+        )
+        assert isinstance(df, pd.DataFrame)
```
Comment (m-richards) [inline geopandas/tests/test_geodataframe.py:790]: if df were to somehow become a GeoDataFrame, isinstance(df, pd.DataFrame) would still be True
```suggestion
        assert type(df) == pd.DataFrame
```

File: geopandas/tests/test_geodataframe.py [hunk 3]
```diff
@@ -772,6 +772,27 @@ def test_dataframe_to_geodataframe(self):
...
+        assert "geometry" not in df
+        pd.testing.assert_frame_equal(df, df_copy)
```
Comment (m-richards) [inline geopandas/tests/test_geodataframe.py:792]: nit, assert_frame_equal is already imported in test_geodataframe
```suggestion
    assert_frame_equal(df, df_copy)
```

**Review by m-richards — APPROVED** [review]
This also looks fine to me, a few small comments.

File: geopandas/tests/test_geodataframe.py (hunk 3 above)
Comment (spolloni) [reply geopandas/tests/test_geodataframe.py]: thanks! changed in https://github.com/geopandas/geopandas/pull/2667/commits/ccfb62a5f4904226c2394ee70f5264bfec9d71ee

File: geopandas/tests/test_geodataframe.py (hunk 2 above)
Comment (spolloni) [reply geopandas/tests/test_geodataframe.py]: good catch, see https://github.com/geopandas/geopandas/pull/2667/commits/ccfb62a5f4904226c2394ee70f5264bfec9d71ee

File: geopandas/tests/test_geodataframe.py (hunk 1 above)
Comment (spolloni) [reply geopandas/tests/test_geodataframe.py:795]: added in in https://github.com/geopandas/geopandas/pull/2667/commits/22c4a91923f89fccf98c70ad64ceb603284fa65d

**Comment by spolloni** [comment]
> Can you also add a changelog note?

@martinfleis done in https://github.com/geopandas/geopandas/pull/2667/commits/30e1ef63abe9f3ccd4e0f7229f4f0d99cf9698d1

**Review by martinfleis — APPROVED** [review]
Thanks! (I have changed the changelog note as there was no dev version section yet.

**Comment by spolloni** [comment]
> as there was no dev version section yet

ah yes, sorry I missed that, thanks!!

**Review by m-richards — APPROVED** [review]

**Comment by martinfleis** [comment]
Thank you @spolloni!

**Pull request merged**